lms/services/blackboard_api.py

```diff
@@ -45,6 +46,32 @@ def get_token(self, authorization_code):
             response.validated_data.get("expires_in"),
         )
 
+    def list_files(self, course_id):
+        """
+        Return the list of files in the given course.
+
+        :raise ProxyAPIAccessTokenError: if we don't have a Blackboard API
+            access token for the current user
+        """
+        response = self._http_service.get(
+            self._api_url(f"courses/uuid:{course_id}/resources"),
+            oauth=True,
+            schema=BlackboardListFilesSchema,
+        )
+
+        return [
+            {
+                "display_name": file_["name"],
+                "id": file_["id"],
+                "updated_at": file_["modified"],
+            }
+            for file_ in response.validated_data
+        ]
```

Comment:
It should be possible for the `list_files()` method to be a one-liner:

```python
return self._http_service.get(
    self._api_url(f"courses/uuid:{course_id}/resources"),
    oauth=True,
    schema=BlackboardListFilesSchema,
)
```

See the comment on the schema below.

Reply:
I decided to go ahead and move the mutation of the data dict (excluding unknown fields and renaming some of the fields) into `BlackboardListFilesSchema`, for a few reasons:

1. It's what Marshmallow schemas are intended for.

   It's not really accurate to think of these as "validation schemas", what they actually are is serialization/deserialization/validation schemas (the name `lms/validation/` may be unfortunate for this reason). If you read the beginning of the Marshmallow docs they describe it as a <q>library for converting complex datatypes, such as objects, to and from native Python datatypes</q> and they actually cover serialization and deserialization first before getting on to validation.

   It might make more sense if you imagine having a `models.File` object and the schema returning a list of those, rather than using dicts directly as the models. Imagine having multiple schemas (`BlackboardFilesSchema`, `CanvasFilesSchema`, `MoodleFilesSchema`) that parse different responses and all return lists of `models.File` objects.

   Validation is a necessity of deserialization because if the input data is invalid then the schema can't create a valid model object from it.

   We already use Marshmallow schemas in this way in the LMS app, e.g. we have schemas for serializing and deserializing bearer tokens and OAuth 2 state params etc.

   I don't think this is unique to Marshmallow: it's typical for schema libraries to both deserialize and validate at once.

2. It makes the code much shorter and simpler.

   You don't really have to add any code to `BlackboardListFilesSchema` to get it to mutate the file dicts in the ways that we want. Excluding unknown fields is just replacing `unknown = INCLUDE` with `EXCLUDE`. Renaming a field is just using `data_key`: `display_name = fields.Str(data_key="name", ...)`.

   Whereas if `BlackboardListFilesSchema` doesn't do the work then `BlackboardAPIClient` has to call `BlackboardListFilesSchema` and then loop over the list of dicts excluding and renaming fields, which is more code.

3. Sort of the same thing as (2) but it makes the tests much simpler. Instead of having `BlackboardListFilesSchema` tests that test that it returns the dicts from Blackboard unmodified and then `BlackboardAPIClient` tests that feed those unmodified dicts into `BlackboardAPIClient` and assert that it modifies them, instead you just have to test that `BlackboardListFilesSchema` returns the modified dicts. The tests for `BlackboardAPIClient` become just `assert returned == HTTPService.get.return_value`

4. It keeps everything about Blackboard Files API responses in one place.

   `BlackboardListFilesSchema` is the one place that has to deal with Blackboard Files API responses. The `BlackboardListFilesSchema` tests are the only tests where fake Blackboard Files responses are needed. If the responsibility is instead spread between `BlackboardListFilesSchema` and `BlackboardAPIClient` then the tests for both need to know about the Blackboard API response format.

5. It makes `BlackboardAPIClient.list_files()` a one-liner